@xen-orchestra/backups/disks/MergeRemoteDisk.mjs

```diff
@@ -0,0 +1,237 @@
+import assert from 'assert'
+import handlerPath from '@xen-orchestra/fs/path'
+import { createLogger } from '@xen-orchestra/log'
+
+import { basename, dirname } from 'path'
+import { Disposable } from 'promise-toolbox'
+import { asyncEach } from '@vates/async-each'
+
+const { warn } = createLogger('remote-disk:merge')
+
+/**
+ * @typedef {Object} MergeState
+ * @property {{ uuid: number }} child
+ * @property {{ uuid: number }} parent
+ * @property {number} currentBlock
+ * @property {number} mergedDataSize
+ * @property {'mergeBlocks' | 'cleanup'} step
+ * @property {number} diskSize
+ */
+
+export class MergeRemoteDisk {
+    /**
+     * @type {MergeState}
+     */
+    #state
+
+    /**
+     * @type {string}
+     */
+    #statePath
+
+    /**
+     * @type {boolean}
+     */
+    #isResuming
+
+    /**
+     * @type {Logger}
+     */
+    #logInfo
+
+    /**
+     * @type {number}
+     */
+    #mergeBlockConcurrency
+
+    /**
+     * @type {Function}
+     */
+    #onProgress
+
+    /**
+     * @type {boolean}
+     */
+    #removeUnused
+
+    /**
+     * @type {number}
+     */
+    #lastStateWrittenAt
+
+    /**
+     * @type {FileAccessor}
+     */
+    #handler
+
+    /**
+     * @param {FileAccessor} handler
+     * @param {Object} params
+     * @param {Function} params.onProgress
+     * @param {Logger} params.logInfo
+     * @param {boolean} params.removeUnused
+     * @param {number} params.mergeBlockConcurrency
+     */
+    constructor(handler, { onProgress = () => { }, logInfo = () => { }, removeUnused = false, mergeBlockConcurrency = 2 }) {
+        this.#handler = handler
+        this.#logInfo = logInfo
+        this.#onProgress = onProgress
+        this.#removeUnused = removeUnused
+        this.#mergeBlockConcurrency = mergeBlockConcurrency
+
+        this.#isResuming = false
+        this.#lastStateWrittenAt = 0
+    }
+
+    /**
+     * @param {RemoteDisk} parentDisk
+     * @param {RemoteDisk} childDisk
+     */
+    async merge(parentDisk, childDisk) {
+        this.#statePath = dirname(parentDisk.getPath()) + '/.' + basename(parentDisk.getPath()) + '.merge.json'
+
+        // Try to load previous state (resume support)
+        try {
+            const mergeStateContent = await this.#handler.readFile(this.#statePath)
+            this.#state = JSON.parse(mergeStateContent)
+            if (this.#state.currentBlock === null) this.#state.currentBlock = 0
```

Comment:
this fix can be removed, it was for an older bug . Also if you keep it, you'll need to keep the coment with it